lib/multi-line/smart-multi-line.c

```diff
@@ -0,0 +1,453 @@
+                  evt_tag_str("filename", sml_file_name));
+    }
+}
+
+static void
+_free_rules(void)
+{
+  for (gint state_ndx = 0; state_ndx < G_N_ELEMENTS(rules_by_from_state); state_ndx++)
+    {
+      if (rules_by_from_state[state_ndx])
+        {
+          g_ptr_array_free(rules_by_from_state[state_ndx], TRUE);
+          rules_by_from_state[state_ndx] = NULL;
+        }
+    }
+
+  for (gint rule_ndx = 0; rule_ndx < rules->len; rule_ndx++)
+    {
+      SmartMultiLineRule *rule = &g_array_index(rules, SmartMultiLineRule, rule_ndx);
+
+      multi_line_pattern_unref(rule->compiled_regexp);
+      g_free(rule->regexp);
+    }
+  g_array_free(rules, TRUE);
+  rules = NULL;
+}
+
+gboolean
+_fsm_transition(SmartMultiLine *self, const gchar *segment, gsize segment_len)
+{
+  GPtrArray *applicable_rules = rules_by_from_state[self->current_state];
+
+  for (gint i = 0; applicable_rules && i < applicable_rules->len; i++)
+    {
+      SmartMultiLineRule *rule = g_ptr_array_index(applicable_rules, i);
+      gboolean match = multi_line_pattern_match(rule->compiled_regexp, (const guchar *) segment, segment_len);
+
+      msg_trace_printf("smart-multi-line: Matching against pattern: %s in state %d, matched %d", rule->regexp,
+                       self->current_state, match);
+      if (match)
+        {
+          self->current_state = rule->to_state;
+          /* the current segment is part of a sequence */
+          return TRUE;
+        }
+    }
+  self->current_state = SMLS_START_STATE;
+  return FALSE;
+}
+
+void
+_process_segment(SmartMultiLine *self, const gchar *segment, gsize segment_len,
+                 gboolean *segment_is_part_of_trace,
+                 gboolean *segment_starts_a_new_trace,
+                 gboolean *segment_ends_trace)
+{
+  *segment_is_part_of_trace = FALSE;
+  *segment_starts_a_new_trace = FALSE;
+  *segment_ends_trace = FALSE;
+
+  gboolean last_segment_ended_the_trace = self->current_state == SMLS_START_STATE;
+
+  gboolean segment_is_trace = _fsm_transition(self, segment, segment_len);
+
+  msg_trace_printf("smart-multi-line: [STEP1] >>%.*s<<, result=%d, state=%d", (int) segment_len, segment,
+                   segment_is_trace, self->current_state);
+  *segment_is_part_of_trace = segment_is_trace;
+
+  if (!(*segment_is_part_of_trace))
+    {
+      /* try again from the start state, the current segment is may be part of a new trace */
+      segment_is_trace = _fsm_transition(self, segment, segment_len);
+
+      msg_trace_printf("smart-multi-line: [STEP2]: >>%.*s<<, result=%d, state=%d", (int) segment_len, segment,
+                       segment_is_trace, self->current_state);
+
+      *segment_is_part_of_trace = segment_is_trace;
+      if (*segment_is_part_of_trace)
+        *segment_starts_a_new_trace = TRUE;
+    }
+  else
+    {
+      if (last_segment_ended_the_trace)
+        *segment_starts_a_new_trace = TRUE;
+      *segment_ends_trace = (self->current_state == SMLS_START_STATE);
+    }
+}
```

Comment:
This function seemed a bit overcomplicated to me (might be a problem with me), so I have reordered it a bit. Things that did not fit for me:
  1. `segment_is_trace` seems to be the same as `*segment_is_part_of_trace`.
  2. There seems to be some `if` + `=`s, which can be simplified to be just `=`.
  3. I think we should allow a rule to start and end a trace at the same time. While it seems weird and I cannot think of any examples of this, I think we should let the user write optionally one-line traces, for example:
```
# dummy trace
start_state,dummy_trace	/dummy trace: optional line: .*/	dummy_trace
start_state,dummy_trace	/dummy trace: mandatory ending with simple conclusion: .*/	start_state
```

The code after my reordering:
```c
void
_process_segment(SmartMultiLine *self, const gchar *segment, gsize segment_len,
                 gboolean *segment_is_part_of_trace,
                 gboolean *segment_starts_a_new_trace,
                 gboolean *segment_ends_trace)
{
  gboolean processing_started_from_start_state = self->current_state == SMLS_START_STATE;

  *segment_is_part_of_trace = _fsm_transition(self, segment, segment_len);
  msg_trace_printf("smart-multi-line: [STEP1] >>%.*s<<, result=%d, state=%d", (int) segment_len, segment,
                   *segment_is_part_of_trace, self->current_state);

  if (*segment_is_part_of_trace)
    {
      *segment_starts_a_new_trace = processing_started_from_start_state;
    }
  else
    {
      /* try again from the start state, the current segment may be part of a new trace */
      *segment_is_part_of_trace = *segment_starts_a_new_trace = _fsm_transition(self, segment, segment_len);
      msg_trace_printf("smart-multi-line: [STEP2]: >>%.*s<<, result=%d, state=%d", (int) segment_len, segment,
                       *segment_is_part_of_trace, self->current_state);
    }

  *segment_ends_trace = *segment_is_part_of_trace && self->current_state == SMLS_START_STATE;
}
```

Reply:
We can leave this as is.